vnstat: port init-script to procd and enable auto-respawn

xdarklight opened the pull request:
Maintainer: @jow- 
Compile tested: LEDE master
Run tested: LEDE master, tested with the Lantiq xrx200 target (BT HomeHub 5A)

Description:
* the first commit updates the vnstat init-script to use procd (which is preparation for the second commit)
* the second commit enables procd's auto-respawn functionality for vnstat to fix a "race condition" between the vnstat startup and setting the correct date/time (which means that traffic stats are lost between system boot and a manual restart of vnstat) - see the commit description for more information

please note that this is the first time that I'm touching anything procd related, so please keep this in mind while reviewing this!

kdarbyshirebryant commented:
I've been battling another race condition related to storing data on an external USB drive that isn't ready/mounted on vnstat startup - https://github.com/openwrt/packages/pull/4682

kdarbyshirebryant commented on net/vnstat/files/vnstat.init:
```diff
@@ -72,6 +72,10 @@ start_service() {
 	procd_set_param stderr 1
 	procd_set_param command /usr/sbin/vnstatd --nodaemon
+	procd_set_param respawn
+	procd_append_param respawn 3600
```
These respawn values are the defaults anyway, so don't need to be specified.

xdarklight replied:
good catch - thanks
I'll fix this once we decide how to proceed with the three separate (partially 
conflicting) pull-requests

xdarklight commented:
unfortunately I think we now have three patches that port vnstat to procd:
* #4615 only ports the init-script to procd but doesn't add new features
* #4565 adds new UCI settings, one of these is no_add
* #4682 hardcodes the no_add flag from @thornley-touchstar pull request, runs vnstat as non-root user and changes START to 95

so all our pull requests are useful, each of them adds a separate puzzle piece.

@thornley-touchstar @kdarbyshirebryant I hope we can work out a plan how to get all of our patches merged!
my proposal how to combine our solutions (feel free to suggest something different) so we don't break git bisect or mix up features with other changes:
1. keep my first patch (porting the init-script to procd)
2. extending the commit message of my second patch (adding respawn) to indicate that it helps with at least one other race condition
3. add the new UCI settings from @thornley-touchstar pull-request and extending the commit message indicating that ``no_add`` can help even further with a race-condition during startup
4. add the non-root user handling from @kdarbyshirebryant 

does this make sense? once we agree on something (which doesn't have to be what I suggested) we need to work out **who** combines our pull requests

kdarbyshirebryant commented:
I've closed my PR - it's the most hackish in terms of using permission 
failure to deal with a race condition, not that I see a way around that.

So that's only 2 PRs to merge.

KDB

On 10/08/17 19:22, Martin Blumenstingl wrote:
> unfortunately I think we now have three patches that port vnstat to procd:
> 
>   * #4615 <https://github.com/openwrt/packages/pull/4615> only ports the
>     init-script to procd but doesn't add new features
>   * #4565 <https://github.com/openwrt/packages/pull/4565> adds new UCI
>     settings, one of these is no_add
>   * #4682 <https://github.com/openwrt/packages/pull/4682> hardcodes the
>     no_add flag from @thornley-touchstar
>     <https://github.com/thornley-touchstar> pull request, runs vnstat as
>     non-root user and changes START to 95
> 
> so all our pull requests are useful, each of them adds a separate puzzle 
> piece.
> 
> @thornley-touchstar <https://github.com/thornley-touchstar> 
> @kdarbyshirebryant <https://github.com/kdarbyshirebryant> I hope we can 
> work out a plan how to get all of our patches merged!
> my proposal how to combine our solutions (feel free to suggest something 
> different) so we don't break git bisect or mix up features with other 
> changes:
> 
>  1. keep my first patch (porting the init-script to procd)
>  2. extending the commit message of my second patch (adding respawn) to
>     indicate that it helps with at least one other race condition
>  3. add the new UCI settings from @thornley-touchstar
>     <https://github.com/thornley-touchstar> pull-request and extending
>     the commit message indicating that |no_add| can help even further
>     with a race-condition during startup
>  4. add the non-root user handling from @kdarbyshirebryant
>     <https://github.com/kdarbyshirebryant>
> 
> does this make sense? once we agree on something (which doesn't have to 
> be what I suggested) we need to work out *who* combines our pull requests
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/openwrt/packages/pull/4615#issuecomment-321633468>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AGMuq7pswiGoKiPlOkryypjIRvYxTcA3ks5sW0pOgaJpZM4OgafY>.
>

xdarklight commented:
@kdarbyshirebryant OK - as I understand your PR the ``no_add`` option from #4565 should also works as a fix for the race condition you are experiencing (because vnstat doesn't auto-create the DB files anymore), right?
we'd still be happy if you could test result of our pull requests (and confirm everything works fine)!

thornley-touchstar commented:
@xdarklight, sounds like a plan! :)

If you maintain the same uci settings, that would be helpful for me as I use a variant of my patch in a production system.

One problem I didn't resolve was the fact that for "vnstat" we need to supply --dbdir (to mount point) or more preferably --config pointing to the generated conf file. I worked around this by modifying /etc/vnstat.conf (DatabaseDir) parameter by hand in my situation. 

I was considering adding a shim script to replace vnstat and then redirect to the real binary with the --config option to the generated conf supplied. I wasn't sure if there was a formal idiom for a shim like this with openwrt/lede (in particular the naming of the original binary file).

I also made it a point to preserve backward compatibility with the "use_uci" config switch/flag.

Let me know if you want me to help with anything, my time is limited, but should be able to spend a bit of time helping with this.

kdarbyshirebryant commented:
On 10/08/17 22:44, Martin Blumenstingl wrote:
> @kdarbyshirebryant <https://github.com/kdarbyshirebryant> OK - as I 
> understand your PR the |no_add| option from #4565 
> <https://github.com/openwrt/packages/pull/4565> should also works as a 
> fix for the race condition you are experiencing (because vnstat doesn't 
> auto-create the DB files anymore), right?

Sadly, no.  '--no-add' is a bit of a red herring and is a 'harmless'(?) 
left over from my earlier attempts to solve.

The scenario I encountered was that sometimes after a reboot, the vnstat 
display in luci would a) lose old data  and b) show additional 
interfaces that I hadn't set to collect (as confirmed in the luci config 
section)  I'd look at the database directory and NOT see any files 
related to these extra interfaces.  Note I have an 8GB USB stick (ext4 
format) which is where I store vnstat dbases.  What was happening:

1) System boots, but has not yet mounted the USB stick.
2) vnstat starts and notes that /mnt/sda1/vnstat contains no 
databases....so it creates dbases from scratch.
3) system now gets around to mounting /mnt/sda1
4) vnstat starts overwriting data in /mnt/sda1/vnstat

The issue is that vnstat has no means to know/determine that an 
externally mounted database directory has actually been mounted by the 
system.  As part of those investigations and to stop vnstat from auto 
adding interfaces that weren't in the config file I added '--no-add' but 
that didn't solve the 'database dir not mounted issue'.

As vnstat runs as root, I couldn't even use a permissions failure on the 
vnstat database directory to indicate mounted/not mounted.  I'm not so 
sure I like vnstat running as root anyway, so the first step was to get 
it to run as a 'vnstat user' - easy as this can be configured in vnstats 
config file.  With that done an umounted '/mnt/sda1/vnstat' doesn't have 
permission to write, whereas a mounted /mnt/sda/vnstat has 775 
vnstat:vnstat.

Unfortunately the original startup script doesn't cope with failure. 
That can be solved by making it run as a service under procd *and* procd 
has a nice 'run as user (and primary group)' feature so doesn't need 
vnstat's config file messing with....other than making sure vnstat user 
has permission to read it.



> we'd still be happy if you could test result of our pull requests (and 
> confirm everything works fine)!

xdarklight commented:
I just rebased this on top of #5855 

please note that this includes the patch from #5855 -> the other PR should be merged first

xdarklight commented:
the commit which enables auto respawn is not needed anymore to fix the race condition during startup since vnstat 1.18:
```
# logread | grep vnstat
Sat Apr 28 08:07:18 2018 daemon.info vnstatd[4378]: Info: vnStat daemon 1.18 started. (pid:4378 uid:0 gid:0)
Sat Apr 28 08:07:18 2018 daemon.info vnstatd[4378]: Info: Latest database update is in the future (db: 2018-04-28 08:39:11 > now: 2018-04-28 08:07:18). Giving the system clock up to 5 minutes to sync before continuing.
Sat Apr 28 08:41:02 2018 daemon.info vnstatd[4378]: Info: Latest database update is no longer in the future (db: 2018-04-28 08:39:11 <= now: 2018-04-28 08:41:02), continuing.
```

I would still keep the commit so vnstat is restarted in case it crashes at some point.
any objections to this?

ldir-EDB0 commented:
None whatsoever - it's a service so should be kept running if possible.

xdarklight commented:
I just updated this pull request:
- rebased against current master (which now already includes the update to vnstat 1.18)
- dropped respawn default values
- updated PKG_RELEASE in the second patch

jow- commented:
We also need an `procd_set_param file /etc/vnstat.conf` to let procd track changes to `vnstat.conf`, otherwise will no reloads will on changes to it.

xdarklight commented:
@jow- thanks for the hint - I fixed (and tested) that
patch 1 in this PR is already updated

xdarklight commented:
I have rebased this PR once again (no functional changes).
is there anything else I need to do before this gets merged?

jow- commented:
No, its all fine. Merged. Thank you!
